integrations/object_store/src/service/mod.rs

```diff
@@ -0,0 +1,385 @@
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use std::fmt::Debug;
+use std::fmt::Formatter;
+use std::sync::Arc;
+
+use object_store::ObjectStore;
+use opendal::raw::oio::MultipartWriter;
+use opendal::raw::*;
+use opendal::Error;
+use opendal::ErrorKind;
+use opendal::*;
+
+mod deleter;
+mod error;
+mod lister;
+mod reader;
+mod writer;
+
+use deleter::ObjectStoreDeleter;
+use error::parse_error;
+use lister::ObjectStoreLister;
+use reader::ObjectStoreReader;
+use writer::ObjectStoreWriter;
+
+/// ObjectStore backend builder
+#[derive(Default)]
+pub struct ObjectStoreBuilder {
+    store: Option<Arc<dyn ObjectStore + 'static>>,
+}
+
+impl Debug for ObjectStoreBuilder {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        let mut d = f.debug_struct("ObjectStoreBuilder");
+        d.finish_non_exhaustive()
+    }
+}
+
+impl ObjectStoreBuilder {
+    /// Set the object store instance
+    pub fn new(store: Arc<dyn ObjectStore + 'static>) -> Self {
+        Self { store: Some(store) }
+    }
+}
+
+impl Builder for ObjectStoreBuilder {
+    type Config = ();
+
+    fn build(self) -> Result<impl Access> {
+        let store = self.store.ok_or_else(|| {
+            Error::new(ErrorKind::ConfigInvalid, "object store is required")
+                .with_context("service", Scheme::Custom("object_store"))
+        })?;
+
+        Ok(ObjectStoreService { store })
+    }
+}
+
+/// ObjectStore backend
+pub struct ObjectStoreService {
+    store: Arc<dyn ObjectStore + 'static>,
+}
+
+impl Debug for ObjectStoreService {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        let mut d = f.debug_struct("ObjectStoreBackend");
+        d.finish_non_exhaustive()
+    }
+}
+
+impl Access for ObjectStoreService {
+    type Reader = ObjectStoreReader;
+    type Writer = MultipartWriter<ObjectStoreWriter>;
+    type Lister = ObjectStoreLister;
+    type Deleter = ObjectStoreDeleter;
+
+    fn info(&self) -> Arc<AccessorInfo> {
+        let info = AccessorInfo::default();
+        info.set_scheme("object_store")
+            .set_root("/")
+            .set_name("object_store")
+            .set_native_capability(Capability {
+                stat: true,
+                stat_with_if_match: true,
+                stat_with_if_unmodified_since: true,
+                read: true,
+                write: true,
+                delete: true,
+                list: true,
+                list_with_limit: true,
+                list_with_start_after: true,
+                ..Default::default()
+            });
+        Arc::new(info)
+    }
+
+    async fn stat(&self, path: &str, _: OpStat) -> Result<RpStat> {
+        let path = object_store::path::Path::from(path);
+        let meta = self.store.head(&path).await.map_err(parse_error)?;
+
+        let mut metadata = Metadata::new(EntryMode::FILE);
```

Comment:
We have seem many silimar patterns, let's extract a util function for the metadata convert.

Reply:
fixed in 8e45f6d4a7